File: Android/app/src/main/java/com/jiny/catchtherule/ui/play/PlayComponents.kt
package com.jiny.catchtherule.ui.play

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.Spring
import androidx.compose.animation.core.spring
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Backspace
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.jiny.catchtherule.R
import com.jiny.catchtherule.core.model.Puzzle
import com.jiny.catchtherule.ui.theme.AppColors
import com.jiny.catchtherule.ui.theme.card

enum class AnswerFeedback { Correct, Wrong }

/** 수열/격자를 카드 셀로 표시. 빈칸은 그라데이션 테두리로 강조. */
@Composable
fun SequenceDisplay(
    puzzle: Puzzle,
    typed: String = "",
    reveal: Boolean = false,
    feedback: AnswerFeedback? = null,
    modifier: Modifier = Modifier,
) {
    val blankText = when {
        reveal -> puzzle.answer
        typed.isEmpty() -> "?"
        else -> typed
    }
    // 정답 시 정답 칸 팝(scale 오버슈트) 애니메이션.
    val pop = remember { Animatable(1f) }
    LaunchedEffect(feedback) {
        if (feedback == AnswerFeedback.Correct) {
            pop.snapTo(1f)
            pop.animateTo(1.22f, spring(dampingRatio = 0.42f, stiffness = Spring.StiffnessMediumLow))
            pop.animateTo(1f, spring(dampingRatio = 0.6f))
        }
    }
    val popScale = pop.value
    val grid = puzzle.grid
    if (puzzle.isPrompt) {
        // 논리형: 질문 문단을 카드로(줄바꿈 보존, 왼쪽 정렬).
        Box(
            modifier.fillMaxWidth().card().padding(20.dp),
        ) {
            Text(
                puzzle.localizedPrompt,
                color = AppColors.TextPrimary,
                fontSize = 18.sp,
                fontWeight = FontWeight.Medium,
                lineHeight = 26.sp,
            )
        }
    } else if (puzzle.isFigure && puzzle.figures != null) {
        FigureNumberRow(figures = puzzle.figures, blankText = blankText, feedback = feedback)
    } else if (puzzle.isFigureSequence) {
        FigureRow(puzzle.figureTokens ?: emptyList(), puzzle, reveal, feedback, popScale)
    } else if (puzzle.type == "equation" && !grid.isNullOrEmpty()) {
        // 수식형: "[2] + [3] = [13]" — 숫자는 박스, 연산자는 사이 텍스트, 빈칸은 강조 박스
        val cols = (grid.maxOfOrNull { it.size } ?: 1).coerceAtLeast(1)
        // 칸 수가 많아도 가로로 넘치지 않게 단계적 축소.
        val fontSize = when { cols >= 9 -> 14.sp; cols >= 8 -> 16.sp; cols >= 7 -> 18.sp; cols >= 6 -> 21.sp; cols >= 5 -> 24.sp; else -> 28.sp }
        val side = (fontSize.value * 1.8f).dp
        androidx.compose.foundation.layout.Column(
            modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(12.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            grid.forEach { row ->
                Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
                    row.forEach { value ->
                        when {
                            value == null -> EqBox(blankText, fontSize, side, true, feedback, popScale)
                            isEqOperator(value) -> Text(
                                value, color = AppColors.TextTertiary, fontSize = fontSize,
                                fontWeight = FontWeight.SemiBold, fontFamily = FontFamily.SansSerif, maxLines = 1, softWrap = false,
                            )
                            else -> EqBox(value, fontSize, side, false, feedback, popScale)
                        }
                    }
                }
            }
        }
    } else if (!grid.isNullOrEmpty()) {
        // 격자형(두 줄/매트릭스/수식형)
        val cols = (grid.maxOfOrNull { it.size } ?: 1).coerceAtLeast(1)
        val spacing = if (cols >= 5) 8.dp else 10.dp
        val fontSize = when { cols >= 5 -> 22.sp; cols >= 4 -> 25.sp; else -> 28.sp }
        val cellHeight = if (grid.size >= 3) 52.dp else 60.dp
        androidx.compose.foundation.layout.Column(
            modifier.fillMaxWidth(),
            verticalArrangement = Arrangement.spacedBy(spacing),
        ) {
            grid.forEach { row ->
                Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(spacing)) {
                    row.forEach { value -> SeqCell(value, blankText, fontSize, cellHeight, feedback, popScale) }
                }
            }
        }
    } else {
        // 단일 행 (칸 수에 따라 폰트/간격/높이 축소 → 스크롤 없이 화면 폭에 맞춤)
        val tokens = puzzle.tokens ?: emptyList()
        val count = tokens.size.coerceAtLeast(1)
        val spacing = when { count >= 7 -> 6.dp; count >= 6 -> 8.dp; count >= 5 -> 10.dp; else -> 12.dp }
        val fontSize = when { count >= 7 -> 20.sp; count >= 6 -> 23.sp; count >= 5 -> 26.sp; else -> 30.sp }
        val cellHeight = if (count >= 6) 66.dp else 76.dp
        Row(
            modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.spacedBy(spacing),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            tokens.forEach { value -> SeqCell(value, blankText, fontSize, cellHeight, feedback, popScale) }
        }
    }
}

@Composable
private fun RowScope.SeqCell(
    value: String?,
    blankText: String,
    fontSize: androidx.compose.ui.unit.TextUnit,
    height: androidx.compose.ui.unit.Dp,
    feedback: AnswerFeedback?,
    popScale: Float = 1f,
) {
    val isBlank = value == null
    val correct = isBlank && feedback == AnswerFeedback.Correct
    val strokeBrush: Brush = when {
        !isBlank -> SolidColor(AppColors.Stroke)
        feedback == AnswerFeedback.Correct -> SolidColor(AppColors.Success)
        feedback == AnswerFeedback.Wrong -> SolidColor(AppColors.Danger)
        else -> AppColors.AccentGradient
    }
    Box(
        Modifier
            .weight(1f)
            .graphicsLayer {
                if (isBlank) { scaleX = popScale; scaleY = popScale }
            }
            .height(height)
            .then(
                if (correct) Modifier.shadow(16.dp, RoundedCornerShape(16.dp),
                    spotColor = AppColors.Success, ambientColor = AppColors.Success)
                else Modifier
            )
            .clip(RoundedCornerShape(16.dp))
            .background(if (isBlank) Color.White.copy(alpha = 0.03f) else AppColors.Card)
            .border(if (isBlank) 2.dp else 1.dp, strokeBrush, RoundedCornerShape(16.dp))
            .padding(horizontal = 2.dp),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = if (isBlank) blankText else (value ?: ""),
            color = if (isBlank) AppColors.TextPrimary else AppColors.TextSecondary,
            fontSize = fontSize,
            fontWeight = FontWeight.Bold,
            fontFamily = FontFamily.SansSerif,
            maxLines = 1,
            softWrap = false,
            textAlign = TextAlign.Center,
        )
    }
}

/** 수식형 숫자/빈칸 박스. */
@Composable
private fun EqBox(
    text: String,
    fontSize: androidx.compose.ui.unit.TextUnit,
    side: androidx.compose.ui.unit.Dp,
    isBlank: Boolean,
    feedback: AnswerFeedback?,
    popScale: Float,
) {
    val strokeBrush: Brush = when {
        !isBlank -> SolidColor(AppColors.Stroke)
        feedback == AnswerFeedback.Correct -> SolidColor(AppColors.Success)
        feedback == AnswerFeedback.Wrong -> SolidColor(AppColors.Danger)
        else -> AppColors.AccentGradient
    }
    Box(
        Modifier
            .graphicsLayer { if (isBlank) { scaleX = popScale; scaleY = popScale } }
            .defaultMinSize(minWidth = side, minHeight = side)
            .clip(RoundedCornerShape(12.dp))
            .background(if (isBlank) Color.White.copy(alpha = 0.03f) else AppColors.Card)
            .border(if (isBlank) 2.dp else 1.dp, strokeBrush, RoundedCornerShape(12.dp))
            .padding(horizontal = 6.dp),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text,
            color = if (isBlank) AppColors.TextPrimary else AppColors.TextSecondary,
            fontSize = fontSize, fontWeight = FontWeight.Bold,
            fontFamily = FontFamily.SansSerif, maxLines = 1, softWrap = false,
        )
    }
}

private fun isEqOperator(s: String): Boolean =
    s in setOf("+", "-", "−", "×", "x", "*", "÷", "/", "=", "·", ">", "<", "→")

/** 정답 시 잠깐 나타나는 "정답!" 배지. */
@Composable
fun CorrectBadge() {
    Row(
        Modifier
            .clip(androidx.compose.foundation.shape.CircleShape)
            .background(AppColors.Success)
            .padding(horizontal = 22.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
    ) {
        Icon(Icons.Filled.CheckCircle, null, tint = Color.White, modifier = Modifier.size(22.dp))
        Text(stringResource(R.string.correct), color = Color.White, fontSize = 20.sp, fontWeight = FontWeight.ExtraBold)
    }
}

@Composable
fun Keypad(
    canSubmit: Boolean,
    onDigit: (Int) -> Unit,
    onDelete: () -> Unit,
    onSubmit: () -> Unit,
) {
    val rows = listOf(
        listOf(1, 2, 3),
        listOf(4, 5, 6),
        listOf(7, 8, 9),
    )
    androidx.compose.foundation.layout.Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        rows.forEach { row ->
            Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                row.forEach { n -> DigitKey(n, Modifier.weight(1f), onDigit) }
            }
        }
        Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
            KeyBox(Modifier.weight(1f), onClick = onDelete) {
                Icon(Icons.AutoMirrored.Filled.Backspace, null, tint = AppColors.TextSecondary, modifier = Modifier.size(22.dp))
            }
            DigitKey(0, Modifier.weight(1f), onDigit)
            Box(
                Modifier
                    .weight(1f)
                    .height(60.dp)
                    .clip(RoundedCornerShape(16.dp))
                    .background(if (canSubmit) AppColors.AccentGradient else SolidColor(Color.White.copy(alpha = 0.08f)))
                    .clickable(enabled = canSubmit) { onSubmit() },
                contentAlignment = Alignment.Center,
            ) {
                Icon(Icons.Filled.Check, null, tint = Color.White, modifier = Modifier.size(24.dp))
            }
        }
    }
}

@Composable
private fun DigitKey(n: Int, modifier: Modifier, onDigit: (Int) -> Unit) {
    KeyBox(modifier, onClick = { onDigit(n) }) {
        Text(n.toString(), color = AppColors.TextPrimary, fontSize = 26.sp, fontWeight = FontWeight.SemiBold)
    }
}

@Composable
private fun KeyBox(modifier: Modifier, onClick: () -> Unit, content: @Composable () -> Unit) {
    Box(
        modifier
            .height(60.dp)
            .card(16.dp)
            .clickable { onClick() },
        contentAlignment = Alignment.Center,
    ) { content() }
}

@Composable
fun ChoicesGrid(choices: List<String>, enabled: Boolean = true, onPick: (String) -> Unit) {
    // 긴 텍스트(논리형 보기)는 폰트를 줄여 한 칸에 들어가게.
    val maxLen = choices.maxOfOrNull { it.length } ?: 1
    val fontSize = when { maxLen >= 12 -> 16.sp; maxLen >= 7 -> 20.sp; else -> 24.sp }
    androidx.compose.foundation.layout.Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        choices.chunked(2).forEach { pair ->
            Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                pair.forEach { c ->
                    Box(
                        Modifier
                            .weight(1f)
                            .height(64.dp)
                            .card(16.dp)
                            .clickable(enabled = enabled) { onPick(c) }
                            .padding(horizontal = 8.dp),
                        contentAlignment = Alignment.Center,
                    ) {
                        Text(
                            c, color = AppColors.TextPrimary, fontSize = fontSize,
                            fontWeight = FontWeight.Bold, maxLines = 2, textAlign = TextAlign.Center,
                        )
                    }
                }
            }
        }
    }
}

/** 도형 보기 4개(2x2). 탭하면 그 도형의 code 를 제출한다. */
@Composable
fun FigureChoicesGrid(
    choices: List<com.jiny.catchtherule.core.model.Figure>,
    enabled: Boolean = true,
    onPick: (String) -> Unit,
) {
    androidx.compose.foundation.layout.Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        choices.chunked(2).forEach { pair ->
            Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                pair.forEach { fig ->
                    Box(
                        Modifier
                            .weight(1f)
                            .height(72.dp)
                            .card(16.dp)
                            .clickable(enabled = enabled) { onPick(fig.code ?: "") },
                        contentAlignment = Alignment.Center,
                    ) {
                        FigureGlyph(figure = fig, size = 46.dp)
                    }
                }
            }
        }
    }
}

File: iOS/CatchTheRule/Features/Play/PlayComponents.swift
import SwiftUI

// MARK: - Feedback state

enum AnswerFeedback: Equatable {
    case correct
    case wrong
}

// MARK: - Sequence display

/// 수열을 카드 셀로 표시. 빈칸(nil)은 그라데이션 테두리로 강조하며
/// 현재 입력값 또는 정답(reveal)을 보여준다.
struct SequenceDisplay: View {
    let puzzle: Puzzle
    var typed: String = ""
    var reveal: Bool = false
    var feedback: AnswerFeedback? = nil
    var onPickStatement: ((Int) -> Void)? = nil   // 모순찾기: 문장 번호(1-based) 탭

    @State private var popScale: CGFloat = 1   // 정답 칸 팝 애니메이션

    var body: some View {
        Group {
            if puzzle.isContradiction {
                contradictionList(puzzle.localizedStatements)
            } else if puzzle.isPrompt {
                promptCard(puzzle.localizedPrompt)
            } else if puzzle.isFigure, let figures = puzzle.figures {
                FigureNumberRow(figures: figures, blankText: blankText(), feedback: feedback)
            } else if puzzle.isFigureSequence {
                figureRow(puzzle.figureTokens ?? [])
            } else if puzzle.type == "equation", let grid = puzzle.grid, !grid.isEmpty {
                equationBody(grid)
            } else if let grid = puzzle.grid, !grid.isEmpty {
                gridBody(grid)
            } else {
                rowBody(puzzle.tokens ?? [])
            }
        }
        .onChange(of: feedback) { _, newValue in
            if newValue == .correct { triggerPop() }
        }
    }

    // MARK: - 모순찾기(문장 목록)

    private static let circledNumbers = ["①","②","③","④","⑤","⑥","⑦","⑧","⑨","⑩","⑪","⑫","⑬","⑭","⑮","⑯","⑰","⑱","⑲","⑳"]

    /// 번호 매긴 문장 카드 목록. 각 카드를 탭하면 그 번호를 제출.
    /// reveal 이면 정답(모순) 문장을 초록, 오답 탭은 빨강으로 강조.
    private func contradictionList(_ statements: [String]) -> some View {
        VStack(spacing: 10) {
            ForEach(Array(statements.enumerated()), id: \.offset) { i, text in
                statementCard(number: i + 1, text: text)
            }
        }
        .frame(maxWidth: .infinity)
    }

    @ViewBuilder
    private func statementCard(number: Int, text: String) -> some View {
        let isAnswer = String(number) == puzzle.answer
        let isPicked = typed == String(number)
        let showCorrect = reveal && isAnswer
        let showWrong = feedback == .wrong && isPicked
        let badge = number <= Self.circledNumbers.count ? Self.circledNumbers[number - 1] : "\(number)"
        Button {
            guard !reveal else { return }
            onPickStatement?(number)
        } label: {
            HStack(alignment: .top, spacing: 12) {
                Text(badge)
                    .font(.system(size: 20, weight: .bold, design: .rounded))
                    .foregroundStyle(showCorrect ? Theme.success : (showWrong ? Theme.danger : Theme.accent2))
                Text(text)
                    .font(.system(size: 16))
                    .foregroundStyle(Theme.textPrimary)
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .fixedSize(horizontal: false, vertical: true)
                if showCorrect {
                    Image(systemName: "exclamationmark.triangle.fill")
                        .font(.system(size: 15))
                        .foregroundStyle(Theme.success)
                }
            }
            .padding(16)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(
                RoundedRectangle(cornerRadius: 16, style: .continuous)
                    .fill(showCorrect ? Theme.success.opacity(0.12) : Theme.card)
            )
            .overlay(
                RoundedRectangle(cornerRadius: 16, style: .continuous)
                    .stroke(showCorrect ? AnyShapeStyle(Theme.success)
                            : (showWrong ? AnyShapeStyle(Theme.danger) : AnyShapeStyle(Theme.stroke)),
                            lineWidth: (showCorrect || showWrong) ? 2 : 1)
            )
        }
        .buttonStyle(.plain)
        .disabled(reveal)
    }

    // MARK: - 논리형(질문 문단)

    /// 논리 퍼즐의 질문 문단을 카드로 표시. 줄바꿈 보존, 왼쪽 정렬.
    private func promptCard(_ text: String) -> some View {
        Text(text)
            .font(.system(size: 18, weight: .medium))
            .foregroundStyle(Theme.textPrimary)
            .lineSpacing(6)
            .frame(maxWidth: .infinity, alignment: .leading)
            .fixedSize(horizontal: false, vertical: true)
            .padding(20)
            .card()
    }

    // MARK: - 도형 시퀀스(시각형)

    /// 도형이 칸별로 변하는 시퀀스. nil 셀이 빈칸(보기에서 고를 도형 자리).
    private func figureRow(_ figs: [Figure?]) -> some View {
        let count = max(1, figs.count)
        let spacing: CGFloat = count >= 6 ? 6 : 10
        let glyph: CGFloat = count >= 6 ? 34 : 42
        return HStack(spacing: spacing) {
            ForEach(Array(figs.enumerated()), id: \.offset) { _, fig in
                if let fig {
                    FigureGlyph(figure: fig, size: glyph)
                        .frame(maxWidth: .infinity)
                        .frame(height: 76)
                        .card(cornerRadius: 16)
                } else {
                    figureBlankCell(glyph: glyph)
                }
            }
        }
        .frame(maxWidth: .infinity)
    }

    // 시퀀스의 빈칸. reveal 이면 정답 도형(code 일치)을 보여준다.
    @ViewBuilder
    private func figureBlankCell(glyph: CGFloat) -> some View {
        let answerFig = reveal ? puzzle.figureChoices?.first(where: { $0.code == puzzle.answer }) : nil
        ZStack {
            if let answerFig {
                FigureGlyph(figure: answerFig, size: glyph)
            } else {
                Text("?")
                    .font(.system(size: 30, weight: .bold, design: .rounded))
                    .foregroundStyle(Theme.textPrimary)
            }
        }
        .frame(maxWidth: .infinity)
        .frame(height: 76)
        .background(RoundedRectangle(cornerRadius: 16, style: .continuous).fill(Color.white.opacity(0.03)))
        .overlay(RoundedRectangle(cornerRadius: 16, style: .continuous)
            .stroke(blankStroke, lineWidth: 2))
        .shadow(color: Theme.success.opacity(feedback == .correct ? 0.75 : 0), radius: 14)
        .scaleEffect(popScale)
    }

    // MARK: - 수식형(type=equation)

    /// 각 줄을 "[2] + [3] = [13]" 처럼 — 숫자는 박스, 연산자는 사이 텍스트, 빈칸은 강조 박스.
    private func equationBody(_ grid: [[String?]]) -> some View {
        let cols = max(1, grid.map(\.count).max() ?? 1)
        // 칸 수가 많아도 가로로 넘치지 않게 폰트/박스를 단계적으로 축소.
        let fontSize: CGFloat = cols >= 9 ? 14 : (cols >= 8 ? 16 : (cols >= 7 ? 18 : (cols >= 6 ? 21 : (cols >= 5 ? 24 : 28))))
        let boxSide: CGFloat = fontSize * 1.8
        return VStack(spacing: 12) {
            ForEach(Array(grid.enumerated()), id: \.offset) { _, row in
                HStack(spacing: 6) {
                    ForEach(Array(row.enumerated()), id: \.offset) { _, value in
                        if value == nil {
                            eqBox(blankText(), fontSize: fontSize, side: boxSide, isBlank: true)
                        } else if isOperator(value!) {
                            Text(value!)
                                .font(.system(size: fontSize, weight: .semibold, design: .rounded))
                                .foregroundStyle(Theme.textTertiary)
                                .frame(minWidth: fontSize * 0.7)
                        } else {
                            eqBox(value!, fontSize: fontSize, side: boxSide, isBlank: false)
                        }
                    }
                }
            }
        }
        .frame(maxWidth: .infinity)
    }

    private func isOperator(_ s: String) -> Bool {
        ["+", "-", "−", "×", "x", "*", "÷", "/", "=", "·", ">", "<", "→"].contains(s)
    }

    /// 수식형 숫자/빈칸 박스.
    private func eqBox(_ text: String, fontSize: CGFloat, side: CGFloat, isBlank: Bool) -> some View {
        Text(text)
            .font(.system(size: fontSize, weight: .bold, design: .rounded))
            .foregroundStyle(isBlank ? Theme.textPrimary : Theme.textSecondary)
            .lineLimit(1)
            .minimumScaleFactor(0.5)
            .frame(minWidth: side, minHeight: side)
            .padding(.horizontal, 6)
            .background(
                RoundedRectangle(cornerRadius: 12, style: .continuous)
                    .fill(isBlank ? Color.white.opacity(0.03) : Theme.card)
            )
            .overlay(
                RoundedRectangle(cornerRadius: 12, style: .continuous)
                    .stroke(isBlank ? blankStroke : AnyShapeStyle(Theme.stroke), lineWidth: isBlank ? 2 : 1)
            )
            .shadow(color: Theme.success.opacity(isBlank && feedback == .correct ? 0.7 : 0), radius: 10)
            .scaleEffect(isBlank ? popScale : 1)
    }

    private func triggerPop() {
        popScale = 1
        withAnimation(.spring(response: 0.22, dampingFraction: 0.42)) { popScale = 1.22 }
        DispatchQueue.main.asyncAfter(deadline: .now() + 0.26) {
            withAnimation(.spring(response: 0.32, dampingFraction: 0.6)) { popScale = 1 }
        }
    }

    /// 단일 행 표시 (칸 수에 따라 폰트/간격 축소 → 스크롤 없이 화면 폭에 맞춤).
    private func rowBody(_ tokens: [String?]) -> some View {
        let count = max(1, tokens.count)
        let spacing: CGFloat = count >= 7 ? 6 : (count >= 6 ? 8 : (count >= 5 ? 10 : 12))
        let fontSize: CGFloat = count >= 7 ? 20 : (count >= 6 ? 23 : (count >= 5 ? 26 : 30))
        let height: CGFloat = count >= 6 ? 66 : 76
        return HStack(spacing: spacing) {
            ForEach(Array(tokens.enumerated()), id: \.offset) { _, value in
                cell(value: value, isBlank: value == nil, fontSize: fontSize, height: height)
            }
        }
        .frame(maxWidth: .infinity)
    }

    /// 격자형(두 줄/매트릭스/수식형) 표시.
    private func gridBody(_ grid: [[String?]]) -> some View {
        let cols = max(1, grid.map(\.count).max() ?? 1)
        let spacing: CGFloat = cols >= 5 ? 8 : 10
        let fontSize: CGFloat = cols >= 5 ? 22 : (cols >= 4 ? 25 : 28)
        let rows = grid.count
        let height: CGFloat = rows >= 3 ? 52 : 60
        return VStack(spacing: spacing) {
            ForEach(Array(grid.enumerated()), id: \.offset) { _, row in
                HStack(spacing: spacing) {
                    ForEach(Array(row.enumerated()), id: \.offset) { _, value in
                        cell(value: value, isBlank: value == nil, fontSize: fontSize, height: height)
                    }
                }
            }
        }
        .frame(maxWidth: .infinity)
    }

    private func blankText() -> String {
        if reveal { return "\(puzzle.answer)" }
        return typed.isEmpty ? "?" : typed
    }

    private var blankStroke: AnyShapeStyle {
        switch feedback {
        case .correct: return AnyShapeStyle(Theme.success)
        case .wrong: return AnyShapeStyle(Theme.danger)
        case .none: return AnyShapeStyle(Theme.accentGradient)
        }
    }

    @ViewBuilder
    private func cell(value: String?, isBlank: Bool, fontSize: CGFloat, height: CGFloat) -> some View {
        Text(isBlank ? blankText() : (value ?? ""))
            .font(.system(size: fontSize, weight: .bold, design: .rounded))
            .lineLimit(1)
            .minimumScaleFactor(0.4)            // 긴 토큰(예: 1211)도 칸 안에 맞춤
            .foregroundStyle(isBlank ? Theme.textPrimary : Theme.textSecondary)
            .frame(maxWidth: .infinity)         // 칸들이 화면 폭을 균등 분할
            .frame(height: height)
            .padding(.horizontal, 2)
            .background(
                RoundedRectangle(cornerRadius: 16, style: .continuous)
                    .fill(isBlank ? Color.white.opacity(0.03) : Theme.card)
            )
            .overlay(
                RoundedRectangle(cornerRadius: 16, style: .continuous)
                    .stroke(isBlank ? blankStroke : AnyShapeStyle(Theme.stroke),
                            lineWidth: isBlank ? 2 : 1)
            )
            .shadow(color: Theme.success.opacity(isBlank && feedback == .correct ? 0.75 : 0),
                    radius: 14)
            .scaleEffect(isBlank ? popScale : 1)
            .zIndex(isBlank ? 1 : 0)
    }
}

// MARK: - Correct badge

/// 정답 시 잠깐 나타나는 "정답!" 배지.
struct CorrectBadge: View {
    var body: some View {
        HStack(spacing: 8) {
            Image(systemName: "checkmark.circle.fill")
            Text(String.loc("correct")).fontWeight(.heavy)
        }
        .font(.system(size: 20, weight: .bold, design: .rounded))
        .foregroundStyle(.white)
        .padding(.horizontal, 22)
        .padding(.vertical, 12)
        .background(Theme.success, in: Capsule())
        .shadow(color: Theme.success.opacity(0.5), radius: 18, y: 6)
    }
}

// MARK: - Keypad

struct KeypadView: View {
    let onDigit: (Int) -> Void
    let onDelete: () -> Void
    let onSubmit: () -> Void
    var canSubmit: Bool

    private let columns = Array(repeating: GridItem(.flexible(), spacing: 12), count: 3)

    var body: some View {
        LazyVGrid(columns: columns, spacing: 12) {
            ForEach(1...9, id: \.self) { n in
                digitKey(n)
            }
            deleteKey
            digitKey(0)
            submitKey
        }
    }

    private func digitKey(_ n: Int) -> some View {
        Button { onDigit(n) } label: {
            Text("\(n)")
                .font(.system(size: 26, weight: .semibold, design: .rounded))
                .foregroundStyle(Theme.textPrimary)
                .frame(maxWidth: .infinity)
                .frame(height: 60)
                .card(cornerRadius: 16)
        }
        .buttonStyle(.plain)
    }

    private var deleteKey: some View {
        Button(action: onDelete) {
            Image(systemName: "delete.left")
                .font(.system(size: 22, weight: .medium))
                .foregroundStyle(Theme.textSecondary)
                .frame(maxWidth: .infinity)
                .frame(height: 60)
                .card(cornerRadius: 16)
        }
        .buttonStyle(.plain)
    }

    private var submitKey: some View {
        Button(action: onSubmit) {
            Image(systemName: "checkmark")
                .font(.system(size: 22, weight: .bold))
                .foregroundStyle(.white)
                .frame(maxWidth: .infinity)
                .frame(height: 60)
                .background {
                    if canSubmit { Theme.accentGradient } else { Color.white.opacity(0.08) }
                }
                .clipShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
        }
        .buttonStyle(.plain)
        .disabled(!canSubmit)
    }
}

// MARK: - Choices

struct ChoicesView: View {
    let choices: [String]
    var disabled: Bool = false
    let onPick: (String) -> Void

    private let columns = Array(repeating: GridItem(.flexible(), spacing: 12), count: 2)

    var body: some View {
        LazyVGrid(columns: columns, spacing: 12) {
            ForEach(choices, id: \.self) { c in
                Button { onPick(c) } label: {
                    Text(c)
                        .font(.system(size: 24, weight: .bold, design: .rounded))
                        .foregroundStyle(Theme.textPrimary)
                        .lineLimit(2)
                        .minimumScaleFactor(0.6)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 8)
                        .frame(maxWidth: .infinity)
                        .frame(height: 64)
                        .card(cornerRadius: 16)
                }
                .buttonStyle(.plain)
            }
        }
        .disabled(disabled)
    }
}

// MARK: - Figure choices (시각형 보기)

/// 도형 보기 4개. 탭하면 그 도형의 code 를 제출한다.
struct FigureChoicesView: View {
    let choices: [Figure]
    var disabled: Bool = false
    let onPick: (String) -> Void

    private let columns = Array(repeating: GridItem(.flexible(), spacing: 12), count: 2)

    var body: some View {
        LazyVGrid(columns: columns, spacing: 12) {
            ForEach(Array(choices.enumerated()), id: \.offset) { _, fig in
                Button { onPick(fig.code ?? "") } label: {
                    FigureGlyph(figure: fig, size: 46)
                        .frame(maxWidth: .infinity)
                        .frame(height: 72)
                        .card(cornerRadius: 16)
                }
                .buttonStyle(.plain)
            }
        }
        .disabled(disabled)
    }
}

// MARK: - Shake effect

struct Shake: GeometryEffect {
    var animatableData: CGFloat
    func effectValue(size: CGSize) -> ProjectionTransform {
        let translation = 10 * sin(animatableData * .pi * 4)
        return ProjectionTransform(CGAffineTransform(translationX: translation, y: 0))
    }
}
